### ISSUE TITLE
Cue/Regular note interaction with beams

### ISSUE THREAD

**Issue by craigsapp** [rism-digital/verovio#808]
When beams contain a mixture of regular and cue-sized notes, note stems are not aligned with notes of the other type:

<img width="866" alt="screen shot 2018-03-29 at 4 13 52 am" src="https://user-images.githubusercontent.com/3487289/38085942-5d5aa650-3307-11e8-8d7e-e8dafa8f3bf8.png">

The first beam has regular notes, and is displayed as expected. The last beam is cue-sized and attaches to cue-sized notes and is displayed as expected.  The second beam is regular, but cue-size notes are not attached to the stems.  The third beam is cue-sized, but regular notes are not attached correctly to the stems.

Also, the last note in the beam group defines whether or not the beam is cue sized.  Perhaps it makes more sense to define the beam's size by the first note of the beam.  Or alternatively, if any note in the beam group is regular size, then the beam is regular size.

MEI test data:

```xml
<?xml version="1.0" encoding="UTF-8"?>
<?xml-model href="http://music-encoding.org/schema/4.0.0/mei-all.rng" type="application/xml" schematypens="http://relaxng.org/ns/structure/1.0"?>
<?xml-model href="http://music-encoding.org/schema/4.0.0/mei-all.rng" type="application/xml" schematypens="http://purl.oclc.org/dsdl/schematron"?>
<mei xmlns="http://www.music-encoding.org/ns/mei" meiversion="4.0.0">
    <meiHead>
        <fileDesc>
            <titleStmt>
                <title />
            </titleStmt>
            <pubStmt />
        </fileDesc>
        <encodingDesc>
            <appInfo>
                <application isodate="2018-03-29T04:12:39" version="2.0.0-dev-6ac488b-dirty">
                    <name>Verovio</name>
                    <p>Transcoded from Humdrum</p>
                </application>
            </appInfo>
        </encodingDesc>
        <workDesc>
            <work>
                <titleStmt>
                    <title />
                </titleStmt>
            </work>
        </workDesc>
        <extMeta>
            <frames xmlns:humxml="http://www.humdrum.org/ns/humxml">
                <metaFrame n="12" token="!!!RDF**kern: @ = cue size" xml:id="loc12">
                    <frameInfo>
                        <startTime float="4" />
                        <frameType>reference</frameType>
                        <referenceKey>RDF**kern</referenceKey>
                        <referenceValue>@ = cue size</referenceValue>
                    </frameInfo>
                </metaFrame>
            </frames>
        </extMeta>
    </meiHead>
    <music>
        <body>
            <mdiv xml:id="mdiv-0000001177490062">
                <score xml:id="score-0000001013664929">
                    <scoreDef xml:id="scoredef-0000000698605973">
                        <staffGrp xml:id="staffgrp-0000002131000993">
                            <staffDef xml:id="staffdef-0000001458054747" clef.shape="G" clef.line="2" n="1" lines="5">
                                <label xml:id="label-0000001191615757" />
                            </staffDef>
                        </staffGrp>
                    </scoreDef>
                    <section xml:id="section-0000000678690052">
                        <measure xml:id="measure-L2">
                            <staff xml:id="staff-L2F1N1" n="1">
                                <layer xml:id="layer-L2F1N1" n="1">
                                    <beam xml:id="beam-L3F1-L4F1">
                                        <note xml:id="note-L3F1" dur="8" oct="4" pname="c" accid.ges="n" />
                                        <note xml:id="note-L4F1" dur="8" oct="4" pname="d" accid.ges="n" />
                                    </beam>
                                    <beam xml:id="beam-L5F1-L6F1">
                                        <note xml:id="note-L5F1" dur="8" oct="4" pname="c" cue="true" accid.ges="n" />
                                        <note xml:id="note-L6F1" dur="8" oct="4" pname="d" accid.ges="n" />
                                    </beam>
                                    <beam xml:id="beam-L7F1-L8F1">
                                        <note xml:id="note-L7F1" dur="8" oct="4" pname="c" accid.ges="n" />
                                        <note xml:id="note-L8F1" dur="8" oct="4" pname="d" cue="true" accid.ges="n" />
                                    </beam>
                                    <beam xml:id="beam-L9F1-L10F1">
                                        <note xml:id="note-L9F1" dur="8" oct="4" pname="c" cue="true" accid.ges="n" />
                                        <note xml:id="note-L10F1" dur="8" oct="4" pname="d" cue="true" accid.ges="n" />
                                    </beam>
                                </layer>
                            </staff>
                        </measure>
                    </section>
                </score>
            </mdiv>
        </body>
    </music>
</mei>
```

**Comment by craigsapp** [comment]
A closely related problem is how to align notes that are cue/regular sized in different layers.

Example:

<img width="639" alt="screen shot 2018-03-29 at 9 53 55 pm" src="https://user-images.githubusercontent.com/3487289/38125043-5682a8e8-339b-11e8-8c65-8cf6d38db7c9.png">

The first measure of the above example is notated correctly, but the second measure has the cue-sized note left-justified to the regular note.  In this case it will have to be right-justified to avoid the stem up on the cue-sized note from appearing to originate in the middle of the regular note (the stem of the cue sized note should be right justified to the regular notehead.

**Comment by rettinghaus** [comment]
Do you have a real-life example for the latter? Haven't seen this in the wild.

**Comment by craigsapp** [comment]
Common in Chopin nocturnes.  For example from page 16 of this PDF:

http://ks.imslp.info/files/imglnks/usimg/0/0c/IMSLP368036-PMLP02312-Chopin_-_Opus_9_(Cortot)_french.pdf

<img width="361" alt="screen shot 2018-03-31 at 3 00 44 am" src="https://user-images.githubusercontent.com/3487289/38162058-5f60fcd6-348f-11e8-8a59-9e31c139d5c7.png">

**Comment by craigsapp** [comment]
Here is the current rendering of the above example in verovio:

<img width="603" alt="screen shot 2018-03-31 at 3 17 54 am" src="https://user-images.githubusercontent.com/3487289/38162187-c2290640-3491-11e8-82ab-94af08d39665.png">

Also notice another improvement could be to make the size of the grace notes and cue-sized notes slightly different.  Notice in the above example that there is a difference in size between the two types of notes.  This is discussed in issue https://github.com/rism-ch/verovio/issues/586 (notes which are both cue-sized and grace notes at the same time would be even smaller at about 50% the size of regular notes).

Data for above example:

```
**kern	**kern
*k[f#c#g#d#a#]	*k[f#c#g#d#a#]
(8FF#L<	4.a#
8BB	.
8D#	.
.	8bqL
.	8a#q
.	8g##q
.	8a#qJ
*	*^
*	*cue	*
*	*rscale:2	*
8BB	(16dd#L	4dd#
.	16ddn	.
8A#	16cc#	.
.	16cc	.
*	*rscale:1	*
8D#J)	8bJ)	8b
*	*Xcue	*
*	*v	*v
=	=
*-	*-
!!!RDF**kern: < = below
```

```xml
<?xml version="1.0" encoding="UTF-8"?>
<?xml-model href="http://music-encoding.org/schema/4.0.0/mei-all.rng" type="application/xml" schematypens="http://relaxng.org/ns/structure/1.0"?>
<?xml-model href="http://music-encoding.org/schema/4.0.0/mei-all.rng" type="application/xml" schematypens="http://purl.oclc.org/dsdl/schematron"?>
<mei xmlns="http://www.music-encoding.org/ns/mei" meiversion="4.0.0">
    <meiHead>
        <fileDesc>
            <titleStmt>
                <title />
            </titleStmt>
            <pubStmt />
        </fileDesc>
        <encodingDesc>
            <appInfo>
                <application isodate="2018-03-31T03:22:08" version="2.0.0-dev-b1351cc">
                    <name>Verovio</name>
                    <p>Transcoded from Humdrum</p>
                </application>
            </appInfo>
        </encodingDesc>
        <workDesc>
            <work>
                <titleStmt>
                    <title />
                </titleStmt>
            </work>
        </workDesc>
        <extMeta>
            <frames xmlns:humxml="http://www.humdrum.org/ns/humxml">
                <metaFrame n="23" token="!!!RDF**kern: &lt; = below" xml:id="loc23">
                    <frameInfo>
                        <startTime float="3" />
                        <frameType>reference</frameType>
                        <referenceKey>RDF**kern</referenceKey>
                        <referenceValue>&lt; = below</referenceValue>
                    </frameInfo>
                </metaFrame>
            </frames>
        </extMeta>
    </meiHead>
    <music>
        <body>
            <mdiv xml:id="mdiv-0000000192177577">
                <score xml:id="score-0000001373573977">
                    <scoreDef xml:id="scoredef-0000001228874379">
                        <staffGrp xml:id="staffgrp-0000000486607978" symbol="brace" barthru="true">
                            <staffDef xml:id="staffdef-0000000234888731" clef.shape="G" clef.line="2" key.sig="5s" n="1" lines="5">
                                <label xml:id="label-0000001660343289" />
                            </staffDef>
                            <staffDef xml:id="staffdef-0000001263545318" clef.shape="F" clef.line="4" key.sig="5s" n="2" lines="5" />
                        </staffGrp>
                    </scoreDef>
                    <section xml:id="section-0000000433128026">
                        <measure xml:id="measure-L1" n="0">
                            <staff xml:id="staff-0000000990397130" n="1">
                                <layer xml:id="layer-L1F2N1" n="1">
                                    <note xml:id="note-L3F2" dots="1" dur="4" oct="4" pname="a" accid.ges="s" />
                                    <beam xml:id="beam-L6F2-L7F2">
                                        <note xml:id="note-L6F2" dur="8" oct="4" pname="b" grace="unacc" accid.ges="n" />
                                        <note xml:id="note-L7F2" dur="8" oct="4" pname="a" grace="unacc" accid.ges="s" />
                                        <note xml:id="note-L8F2" dur="8" oct="4" pname="g" grace="unacc" accid="x" />
                                        <note xml:id="note-L9F2" dur="8" oct="4" pname="a" grace="unacc" accid.ges="s" />
                                    </beam>
                                    <beam xml:id="beam-L14F2-L19F2">
                                        <note xml:id="note-L14F2" dur.ges="16" dur="8" oct="5" pname="d" cue="true" accid.ges="s" />
                                        <note xml:id="note-L15F2" dur.ges="16" dur="8" oct="5" pname="d" cue="true" accid="n" />
                                        <note xml:id="note-L16F2" dur.ges="16" dur="8" oct="5" pname="c" cue="true" accid.ges="s" />
                                        <note xml:id="note-L17F2" dur.ges="16" dur="8" oct="5" pname="c" cue="true" accid="n" />
                                        <note xml:id="note-L19F2" dur="8" oct="4" pname="b" cue="true" accid.ges="n" />
                                    </beam>
                                </layer>
                                <layer xml:id="layer-L11F3N2" n="2">
                                    <space xml:id="space-0000000948600237" dots="1" dur="4" />
                                    <note xml:id="note-L14F3" dur="4" oct="5" pname="d" accid.ges="s" />
                                    <note xml:id="note-L19F3" dur="8" oct="4" pname="b" accid.ges="n" />
                                </layer>
                            </staff>
                            <staff xml:id="staff-0000001539952117" n="2">
                                <layer xml:id="layer-L1F1N1" n="1">
                                    <beam xml:id="beam-L3F1-L19F1">
                                        <note xml:id="note-L3F1" dur="8" oct="2" pname="f" stem.dir="down" accid.ges="s" />
                                        <note xml:id="note-L4F1" dur="8" oct="2" pname="b" stem.dir="down" accid.ges="n" />
                                        <note xml:id="note-L5F1" dur="8" oct="3" pname="d" stem.dir="down" accid.ges="s" />
                                        <note xml:id="note-L14F1" dur="8" oct="2" pname="b" stem.dir="down" accid.ges="n" />
                                        <note xml:id="note-L16F1" dur="8" oct="3" pname="a" stem.dir="down" accid.ges="s" />
                                        <note xml:id="note-L19F1" dur="8" oct="3" pname="d" stem.dir="down" accid.ges="s" />
                                    </beam>
                                </layer>
                            </staff>
                            <slur xml:id="slur-L14F2-L19F2N1" staff="1" startid="#note-L14F2" endid="#note-L19F2" />
                            <slur xml:id="slur-L3F1-L19F1N1" staff="2" startid="#note-L3F1" endid="#note-L19F1" />
                        </measure>
                    </section>
                </score>
            </mdiv>
        </body>
    </music>
</mei>
```

**Comment by rettinghaus** [comment]
Chopin, of course!

**Comment by lpugin** [comment]
> Also, the last note in the beam group defines whether or not the beam is cue sized. Perhaps it makes more sense to define the beam's size by the first note of the beam. Or alternatively, if any note in the beam group is regular size, then the beam is regular size.

Yes, picking the last or the first note is not really satisfactory. It would be useful to have a `@cue` on `<beam>`, and then with mixed sizes in a beam the size would be regular unless specified differently with its `@cue`. (That would also be usable with cues unissons in regular size.)

**Comment by craigsapp** [comment]
Yes, allowing for `beam@cue` would be useful for completeness.  I think I have seen regular and cue notes displayed under the same beam (for realizing ornaments), although I cannot think of a specific real example.

**Comment by lpugin** [comment]
@pe-ro? What do you think?

**Comment by pe-ro** [comment]
Adding `@cue` to `<beam>` is fine with me.  The following modification --

```
<classSpec ident="att.beam.vis" module="MEI.visual" type="atts">
    <desc>Visual domain attributes.</desc>
    <classes>
      <memberOf key="att.color"/>
      <memberOf key="att.cue"/>
      <memberOf key="att.beamRend"/>
    </classes>
  </classSpec>
```
does the job.

Can I get a feature request please?

You want it now or in version 5.0.0?

**Comment by craigsapp** [comment]
Cue-sized beam control is implemented in the MEI 4.0.0 ODD specification: https://github.com/music-encoding/music-encoding/commit/8507516acfe8b07b37eb15e7de9721b1781e74c8

The following rules should probably control selection of cue/regular beams:

(1) If there is a `beam@cue` attribute, use it
(2) If all of the notes in the beam are of one type, then use the same type of beam (the current method)
(3) If there are a mix of cue-sized and regular-sized notes sharing a beam, then use a regular-sized beam by default (which can be changed by setting `beam@cue="true"`).
